File: aerox.Runtime.Widgets/Widget.cs
using System.Runtime.InteropServices;
using aerox.Runtime.Math;
using aerox.Runtime.Widgets.Events;
using aerox.Runtime.Widgets.Graphics;

namespace aerox.Runtime.Widgets;

// [StructLayout(LayoutKind.Sequential)]
// public struct WidgetPushConstants
// {
//     public Vector4<float> clip;
//
//     public Vector4<float> extent;
// }
[StructLayout(LayoutKind.Sequential)]
public struct WidgetPushConstants
{
    public Matrix3 Transform;

    public Vector2<float> Size;
}

public abstract class Widget : Disposable
{
    private Size2d? _cachedDesiredSize;

    private Surface? _cursorUpRoot;
    private Vector2<float> _relativeOffset = 0.0f;
    private Size2d _size = new();
    private SWidgetsModule _subsystem = SRuntime.Get().GetModule<SWidgetsModule>();

    /// <summary>
    ///     The local angle to render this widget at
    /// </summary>
    public float Angle = 0.0f;

    

    /// <summary>
    ///     The pivot used to render this widget. Affects <see cref="Angle" /> and <see cref="Scale" />.
    /// </summary>
    public Vector2<float> Pivot = 0.0f;

    /// <summary>
    ///     The local scale to apply to this widget
    /// </summary>
    public Vector2<float> Scale = 1.0f;

    public bool Hovered { get; private set; }
    public Surface? Surface { get; private set; }

    public Container? Parent { get; private set; }

    public WidgetVisibility Visibility { get; set; } = WidgetVisibility.Visible;

    private WidgetPadding _padding = new WidgetPadding();
    /// <summary>
    ///     The Padding For This Widget (Left, Top, Right, Bottom)
    /// </summary>
    public WidgetPadding Padding
    {
        get => _padding;
        set
        {
            _padding = value;
            CheckSize();
        }
    }


    protected bool IsPendingMouseUp => _cursorUpRoot is { Disposed: false };
    
    /// <summary>
    /// Check if this widget is focused by its current surface
    /// </summary>
    public bool Focused => Surface?.FocusedWidget == this;


    /// <summary>
    ///     Computes the relative/local transformation matrix for this widget
    /// </summary>
    /// <returns></returns>
    public Matrix3 ComputeRelativeTransform() =>
        Matrix3.Identity.Translate(GetOffset()).RotateDeg(Angle).Translate(GetSize() * Pivot - 1.0f);


    public Matrix3 ComputeAbsoluteTransform()
    {
        var parentTransform = Parent?.ComputeAbsoluteTransform() ?? Matrix3.Identity;
        return ComputeRelativeTransform() * parentTransform;
    }

    public void SetVisibility(WidgetVisibility visibility)
    {
        Visibility = visibility;
    }

    public Rect GetRect()
    {
        return new Rect(_relativeOffset, _size);
    }

    public Rect GetRect(Vector2<float> offset)
    {
        return new Rect(offset + _relativeOffset, _size);
    }

    public void SetParent(Container? widget)
    {
        Parent = widget;
    }

    public bool IsSelfHitTestable()
    {
        return Visibility is WidgetVisibility.Visible or WidgetVisibility.VisibleNoHitTestSelf;
    }

    public bool IsChildrenHitTestable()
    {
        return Visibility is WidgetVisibility.Visible or WidgetVisibility.VisibleNoHitTestChildren;
    }

    public bool IsHitTestable()
    {
        return IsSelfHitTestable() || IsChildrenHitTestable();
    }


    public virtual void NotifyAddedToSurface(Surface surface)
    {
        Surface = surface;
        OnAddedToSurface(surface);
    }

    public virtual void NotifyRemovedFromSurface(Surface surface)
    {
        if (Surface == surface) Surface = null;
        OnRemovedFromSurface(surface);
    }

    protected virtual void OnAddedToSurface(Surface surface)
    {
    }

    protected virtual void OnRemovedFromSurface(Surface surface)
    {
    }


    public virtual Widget? ReceiveCursorDown(CursorDownEvent e, TransformInfo info)
    {
        if (IsSelfHitTestable())
            if (OnCursorDown(e))
            {
                _cursorUpRoot = Surface;
                BindCursorUp();
                return this;
            }

        return null;
    }

    protected void BindCursorUp()
    {
        var root = _cursorUpRoot;
        if (root?.Disposed == false) root.OnCursorUp += ReceiveCursorUp;
    }

    protected void UnBindCursorUp()
    {
        var root = _cursorUpRoot;
        if (root?.Disposed == false) root.OnCursorUp -= ReceiveCursorUp;

        _cursorUpRoot = null;
    }

    public virtual void ReceiveCursorUp(CursorUpEvent e)
    {
        UnBindCursorUp();
        OnCursorUp(e);
    }

    public virtual void ReceiveCursorEnter(CursorMoveEvent e, TransformInfo info, List<Widget> items)
    {
        if (!IsSelfHitTestable()) return;
        items.Add(this);
        if (Hovered) return;

        Hovered = true;
        OnCursorEnter(e);
    }

    public virtual bool ReceiveCursorMove(CursorMoveEvent e, TransformInfo info)
    {
        if (IsSelfHitTestable() && OnCursorMove(e)) return true;

        return false;
    }

    public virtual bool ReceiveScroll(ScrollEvent e, TransformInfo info)
    {
        return IsSelfHitTestable() && OnScroll(e);
    }

    protected virtual bool OnCursorDown(CursorDownEvent e)
    {
        return false;
    }

    public virtual void OnCursorUp(CursorUpEvent e)
    {
    }

    protected virtual bool OnCursorMove(CursorMoveEvent e)
    {
        return false;
    }

    protected virtual void OnCursorEnter(CursorMoveEvent e)
    {
    }

    public virtual void ReceiveCursorLeave(CursorMoveEvent e)
    {
        if (!Hovered) return;

        Hovered = false;
        OnCursorLeave(e);
    }

    protected virtual void OnCursorLeave(CursorMoveEvent e)
    {
    }

    protected virtual bool OnScroll(ScrollEvent e)
    {
        return false;
    }
    
    public virtual void OnFocus()
    {
    }
    
    public virtual void OnFocusLost()
    {
    }

    protected override void OnDispose(bool isManual)
    {
        UnBindCursorUp();
    }

    public Vector2<float> GetOffset()
    {
        return _relativeOffset;
    }

    public void SetOffset(Vector2<float> offset)
    {
        _relativeOffset = offset;
    }

    public Size2d GetContentSize()
    {
        return _size - new Vector2<float>(Padding.Left + Padding.Right, Padding.Top + Padding.Bottom);
    }

    public virtual void SetSize(Size2d size)
    {
        _size = size;
    }
    
    public Size2d GetSize()
    {
        return _size;
    }

    public Size2d GetDesiredSize()
    {
        return _cachedDesiredSize ??= ComputeFinalDesiredSize();
    }
    
    public Size2d GetDesiredContentSize()
    {
        return GetDesiredSize() - new Vector2<float>(Padding.Left + Padding.Right, Padding.Top + Padding.Bottom);
    }

    protected virtual Size2d ComputeDesiredSize() => ComputeContentDesiredSize() +
                                                     new Vector2<float>(Padding.Left + Padding.Right,
                                                         Padding.Top + Padding.Bottom);
    
    protected abstract Size2d ComputeContentDesiredSize();

    private Size2d ComputeFinalDesiredSize() =>
        ComputeDesiredSize() + new Vector2<float>(Padding.Left + Padding.Right, Padding.Top + Padding.Bottom);

    protected virtual bool CheckSize()
    {
        var newSize = ComputeFinalDesiredSize();

        if (newSize.Equals(_cachedDesiredSize)) return false;

        _cachedDesiredSize = newSize;
        Parent?.OnSlotUpdated(this);

        return true;
    }
    
    /// <summary>
    /// Computes the transform info of content 
    /// </summary>
    /// <param name="widget">The content</param>
    /// <param name="info">The Absolute Transform info of this widget</param>
    /// <param name="withPadding">Should we also account for padding ? (should be true except when used in <see cref="CollectContent"/>)</param>
    /// <returns>The Absolute Transform info of content</returns>
    public virtual TransformInfo OffsetTransformTo(Widget widget, TransformInfo info, bool withPadding = true)
    {
        var newTransform = withPadding
            ? info.Transform.Translate(new Vector2<float>(Padding.Left, Padding.Top))
            : info.Transform;
        
        return new TransformInfo(newTransform * widget.ComputeRelativeTransform(),widget.GetSize(),info.Depth + 1);
    }


    public virtual void Collect(TransformInfo info,DrawCommands drawCommands)
    {
        if (Visibility is WidgetVisibility.Hidden or WidgetVisibility.Collapsed)
        {
            return;
        }
        
        CollectContent(new TransformInfo(info.Transform.Translate(new Vector2<float>(Padding.Left,Padding.Top)),GetContentSize(),info.Depth),drawCommands);
    }
    
    public abstract void CollectContent(TransformInfo info,DrawCommands drawCommands);
}

File: modules/widgets/lib/rin/widgets/Widget.cpp
#include "rin/widgets/Widget.hpp"
#include "rin/widgets/ContainerWidget.hpp"
#include "rin/widgets/WidgetSurface.hpp"
#include "rin/widgets/event/CursorDownEvent.hpp"

void Widget::OnAddedToSurface(const Shared<WidgetSurface>& widgetSurface)
{
    _surface = widgetSurface;
}

void Widget::OnRemovedFromSurface(const Shared<WidgetSurface>& widgetSurface)
{
    _surface = {};
}

bool Widget::OnCursorDown(const Shared<CursorDownEvent>& event)
{
    return false;
}

void Widget::OnCursorUp(const Shared<CursorUpEvent>& event)
{
}

bool Widget::OnCursorMove(const Shared<CursorMoveEvent>& event)
{
    return false;
}

void Widget::OnCursorEnter(const Shared<CursorMoveEvent>& event)
{
}

void Widget::OnCursorLeave(const Shared<CursorMoveEvent>& event)
{
}

bool Widget::OnScroll(const Shared<ScrollEvent>& event)
{
    return false;
}

void Widget::OnFocus()
{
}

void Widget::OnFocusLost()
{
}

Vec2<float> Widget::ComputeDesiredSize()
{
    const auto padding = GetPadding();
    return ComputeContentSize() + Vec2<float>{padding.left + padding.right, +padding.top + padding.bottom};
}

Shared<WidgetSurface> Widget::GetSurface() const
{
    if (auto surface = _surface.lock())
    {
        return surface;
    }

    return {};
}

Shared<ContainerWidget> Widget::GetParent() const
{
    if (auto parent = _parent.lock())
    {
        return parent;
    }

    return {};
}

Matrix3<float> Widget::ComputeRelativeTransform() const
{
    return Matrix3<float>(1.0f).Translate(GetOffset()).RotateDeg(angle).Translate(GetSize() * pivot * -1.0f);
}

Matrix3<float> Widget::ComputeAbsoluteTransform() const
{
    if (auto parent = GetParent())
    {
        return ComputeRelativeTransform() * parent->ComputeAbsoluteTransform();
    }
    return ComputeRelativeTransform();
}

void Widget::SetVisibility(WidgetVisibility visibility)
{
    auto oldVisibility = _visibility;
    _visibility = visibility;

    if (oldVisibility != _visibility)
    {
        if (_visibility == WidgetVisibility::Collapsed)
        {
            CheckSize();
        }
    }
}

WidgetVisibility Widget::GetVisibility() const
{
    return _visibility;
}

WidgetRect Widget::GetRect() const
{
    return WidgetRect{_offset, _size};
}


void Widget::SetParent(const Shared<ContainerWidget>& container)
{
    _parent = container;
}

void Widget::SetSurface(const Shared<WidgetSurface>& surface)
{
    _surface = surface;
}

bool Widget::IsHovered() const
{
    return _hovered;
}

bool Widget::IsHitTestable() const
{
    return IsSelfHitTestable() || AreChildrenHitTestable();
}

bool Widget::IsSelfHitTestable() const
{
    return _visibility == WidgetVisibility::Visible || _visibility == WidgetVisibility::VisibleNoHitTestChildren;
}

bool Widget::AreChildrenHitTestable() const
{
    return _visibility == WidgetVisibility::Visible || _visibility == WidgetVisibility::VisibleNoHitTestSelf;
}

bool Widget::HasSurface() const
{
    return static_cast<bool>(GetSurface());
}

bool Widget::HasParent() const
{
    return static_cast<bool>(GetParent());
}

void Widget::BindCursorUp(const Shared<WidgetSurface>& surface)
{
    _cursorUpBinding.UnBind();
    _cursorUpBinding = surface->onCursorUp->Add(this->GetSharedDynamic<Widget>(), &Widget::OnCursorUp);
}

void Widget::UnBindCursorUp()
{
    _cursorUpBinding.UnBind();
}

void Widget::NotifyAddedToSurface(const Shared<WidgetSurface>& widgetSurface)
{
    _surface = widgetSurface;
    OnAddedToSurface(widgetSurface);
}

void Widget::NotifyRemovedFromSurface(const Shared<WidgetSurface>& widgetSurface)
{
    _surface = {};
    OnRemovedFromSurface(widgetSurface);
}

Shared<Widget> Widget::NotifyCursorDown(const Shared<CursorDownEvent>& event, const TransformInfo& transform)
{
    if (IsSelfHitTestable() && OnCursorDown(event))
    {
        BindCursorUp(event->surface);
        return this->GetSharedDynamic<Widget>();
    }
    return {};
}

void Widget::NotifyCursorUp(const Shared<CursorUpEvent>& event)
{
    UnBindCursorUp();
    OnCursorUp(event);
}

void Widget::NotifyCursorEnter(const Shared<CursorMoveEvent>& event, const TransformInfo& transform,
                               std::vector<Shared<Widget>>& items)
{
    if (!IsSelfHitTestable()) return;
    items.push_back(this->GetSharedDynamic<Widget>());
    if (IsHovered()) return;

    _hovered = true;
    OnCursorEnter(event);
}

bool Widget::NotifyCursorMove(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
{
    if (IsSelfHitTestable() && OnCursorMove(event)) return true;

    return false;
}

bool Widget::NotifyScroll(const Shared<ScrollEvent>& event, const TransformInfo& transform)
{
    return IsSelfHitTestable() && OnScroll(event);
}

void Widget::NotifyCursorLeave(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
{
    if (!IsHovered()) return;

    _hovered = false;
    OnCursorLeave(event);
}

Vec2<float> Widget::GetOffset() const
{
    return _offset;
}

void Widget::SetOffset(const Vec2<float>& offset)
{
    _offset = offset;
}

Vec2<float> Widget::GetSize() const
{
    return _size;
}

Vec2<float> Widget::GetContentSize() const
{
    auto padding = GetPadding();
    return _size - Vec2{padding.left + padding.right, padding.top + padding.bottom};
}

void Widget::SetSize(const Vec2<float>& size)
{
    _size = size;
}

Vec2<float> Widget::GetDesiredSize()
{
    if (_cachedDesiredSize.has_value())
    {
        return _cachedDesiredSize.value();
    }

    _cachedDesiredSize = ComputeDesiredSize();

    return _cachedDesiredSize.value();
}

Padding Widget::GetPadding() const
{
    return _padding;
}

void Widget::SetPadding(const Padding& padding)
{
    _padding = padding;
    CheckSize();
}

std::optional<Vec2<float>> Widget::GetCachedDesiredSize() const
{
    return _cachedDesiredSize;
}

bool Widget::CheckSize()
{
    auto newSize = ComputeDesiredSize();
    if (_cachedDesiredSize.has_value() && _cachedDesiredSize->NearlyEquals(newSize, 0.01f)) return false;

    _cachedDesiredSize = newSize;

    if (auto parent = GetParent())
    {
        parent->OnChildResized(this);
    }

    return true;
}

void Widget::Collect(const TransformInfo& transform, WidgetDrawCommands& drawCommands)
{
    auto visibility = GetVisibility();
    if (visibility == WidgetVisibility::Hidden || visibility == WidgetVisibility::Collapsed)
    {
        return;
    }

    auto padding = GetPadding();
    CollectContent(TransformInfo{transform.transform.Translate(Vec2{padding.left,padding.top}),transform.size,transform.depth}, drawCommands);
}
